Divider: Adding initial implementation of converged component (#16616)
microsoft/fluentui: packages/react-divider/Spec.md

# Divider Spec

## Background

> The divider control is to replace the Separator control currently in use. Its purpose is to visual separate content.

Author: Updated

> Adding minor structural updates and new engineering practices and update the naming conventions to match naming conventions to OpenUI.

Author: updated

## Prior Art

> [Github Epic Convergance Issue #16254](https://github.com/microsoft/fluentui/issues/16254)

Author: updated

> [Github Convergance Issue #15759](https://github.com/microsoft/fluentui/issues/15759)

Author: updated

[Fluent UI Separator](https://developer.microsoft.com/en-us/fluentui#/controls/web/separator)

[Open UI Component Matrix](https://open-ui.org/analysis/component-matrix)

[Ant Design Divider](https://ant.design/components/divider/)

[Fast Divider](https://explore.fast.design/components/fast-divider)

[Semantic UI Divider](https://semantic-ui.com/elements/divider.html)

## Comparison of v8 and v0

- v8 - [Separator](https://developer.microsoft.com/en-us/fluentui#/controls/web/separator)
- v0 - [Divider](https://fluentsite.z22.web.core.windows.net/0.51.4/components/divider/definition)

## Sample Code

Basic Examples:

```html
<Divider />

> <Divider vertical="true" />

Author: updated

<Divider>This is a divider with content</Divider>
```

## Variants

> The divider will have two main variants, a horzontal and vertical render. The control is the same with the optional property `vertical="true"` defining the rendering type. As per current definition, content can only be added if the divider is horizontal (TBD).

Author: No longer applicable

Standard default horizontal

```html
<Divider />
```

Vertical

```html

> <Divider vertical="true" />

Author: updated

```

## API

From [Divider.types.tsx](http://about:blank)

### Slots

- content (?)

### Props

```ts
export type DividerProps = ComponentProps &
  React.HTMLAttributes<HTMLElement> & {
    /**

>      * A divider can justify the content. Center is default.

Author: updated

     */
    alignContent?: 'start' | 'end' | 'center';

>     children?: any;

Author: updated

    /* A divider can have a overriding color */

>     color?: string;

Reviewer: I think this should be a token since it only affects styling and nothing else.

Author: As per discussion this is will be left in until we have a token story

    /**
     * Shorthand children content within the button.
     */

>     content?: ShorthandProps;

Reviewer: Should we have both `content` and `children`?

Author: updated (removed content)

>     /* A divider can remove any extra margins before or after the element */

Reviewer: nit:
```suggestion
    /* A divider can remove any extra margins before or after the element. */
```

Author: no longer applicable

    fitted?: boolean;

>     /* important will emphasis the content */

Reviewer: ```suggestion
    /* A divider can be classified as important to emphasize its content. */
```

Author: updated

    important?: boolean;

    /* The size is a multiplier value for the size */

>     size?: SizeValue;

Reviewer: I think this description is a bit confusing. Also, will `SizeValue` be a ramp of values? If so, what are the values that it can take?

Author: no longer applicable

    tokens?: RecursivePartial<DividerTokens>;

>     /* A divider can be horizontal (default) or verticle*/

Reviewer: nit:
```suggestion
    /* A divider can be horizontal (default) or vertical. */
```

Author: updated

    vertical?: boolean;
  };
```

## Styling Tokens

TBD

```ts
export type DividerTokens = ColorTokens & FontTokens & {};
```

## Structure

<div class="root">

>   <!-- ::before to handle the divider line for with or without content -->

Reviewer: nit:
```suggestion
  <!-- ::before to handle the divider line independent of the divider having content or not. -->
```

Author: updated

  <div>Content</div>

>   <!-- ::after to handle the post divider line if content exists -->

Reviewer: nit:
```suggestion
  <!-- ::after to handle the divider line that goes after the content if it has been provided. -->
```

Author: updated

</div>

## Migration

The divider control is a direct replacement of the current Separator control, all current properties will remain intact, however the tag name will need to be updated from `<Separator/>` to `<Divider/>`

Migration from V0 will require removing the content property and nesting that content as a child of the node ex.

```
<Divider content="sdfdsF">
<Divider>sfdjkhjsdfhjsdf</Divider>

> ```

Reviewer: This seems to be different from the API and structure you outline above. What is the approach that we are taking?

Author: This was tbd .. has been updated to reflect accordingly

## Behaviors

This component has no state.
This control will have no interactions.

> This control will have no affect on screen readers.

Reviewer: nit:
```suggestion
This control will have no effect on screen readers.
```

Author: updated

> Content if provided, will self determine it's behaviors.

Reviewer: nit:
```suggestion
Content, if provided, will self determine it's behaviors.
```

Author: updated

## Accessibility

> Since the divider is not a focusable element itself, there doesn't need to have an accessilibity role for this control. The descendant content when wrapped from the divider must handle any accessibility information itself when appropriate.

Reviewer: ```suggestion
Since the divider is not a focusable element itself, this control doesn't require any accessibility roles or special behavior applied to it. The descendant content when wrapped from the divider must handle any required accessibility behaviors itself when appropriate.
```

Author: updated

Basic Examples:

```html
<Divider />
<Divider vertical />
<Divider>This is a divider with content</Divider>
```

## Variants

The divider will have two main variants, rendering horizontally or vertically. The control is the same with the optional property `vertical` defining how it renders.

Standard default horizontal

```html
<Divider />
```

Vertical

```html
<Divider vertical />
```

Color

```html
<Divider color="black" />
```

Content

```html
<Divider>My Content</Divider>
```

Icon

```html
<Divider><Icon /></Divider>
```

Important

```html
<Divider important>This is important!</Divider>
```

Inset

```html
<Divider inset />
```

Appearance

```html
<Divider apperance="subtle" />
<Divider apperance="brand" />
<Divider apperance="strong" />
```

## API

From [Divider.types.tsx](https://github.com/microsoft/fluentui/blob/master/packages/react-divider/src/components/Divider/Divider.types.ts)

### Slots

- root

### Props

```ts
export interface DividerProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  /**
   * Determines the alignment of the content within the divider.
   * @defaultvalue 'center'
   */
  alignContent?: 'start' | 'end' | 'center';

>   appearance?: 'default' | 'subtle' | 'brand' | 'strong';

Reviewer 2: Add a comment for `appearance`.

Also, for any prop that has a default value, add a `@defaultvalue` annotation to the comment.

Author: Updated

  children?: any;

  /* A divider can have a overriding border color */

>   color?: string;

Author: Exposing color was meant as a compromise between internal and external usage of the components with the default functionality to be the token story.

>   /* A divider can be classified as important to emphasize its content */

Author: Updated

  important?: boolean;

>   inset?: boolean;

Author: adding comment to comment

  /* A divider can be horizontal (default) or vertical*/
  vertical?: boolean;

  /* Overrides for border visuals */
  borderStyle?: string;
  borderSize?: string | number;

  /* Exposed used properties */
  /*
   *** NOTE ***
    Once we have a full story and patterns for component tokens,
    the following should be removed and the pattern updated.
  */
  fontColor?: string;
  fontSize?: string;
  fontWeight?: string;
  height?: string;
  margin?: string;
  marginBottom?: string;
  marginLeft?: string;
  marginRight?: string;
  marginTop?: string;

>   width?: string;

Author: Removed from Spec.md

}
```

## Styling Tokens

The current exposed tokens and default values are listed below:

**Default**

Font Color : tokens.alias.color.neutral.neutralForeground2

Border Color: tokens.alias.color.neutral.neutralStroke2

**Subtle**

Font Color : tokens.alias.color.neutral.neutralForeground2

Border Color: tokens.alias.color.neutral.neutralStroke3

**Brand**

Font Color : tokens.alias.color.brand.brandBackgroundStatic

Border Color: tokens.alias.color.brand.brandBackgroundStatic

**Brand**

Font Color : tokens.alias.color.neutral.neutralForeground2

Border Color: tokens.alias.color.neutral.neutralStroke1

## Structure

```
<div>
  <!-- ::before to handle the divider line independent of the divider having content or not -->
  <div>Content</div>
  <!-- ::after to handle the divider line that goes after the content if it has been provided. -->
</div>
```

## Migration

The divider control is a direct replacement of the current Separator control. Notable changes are listed below:

> Migration from V0 will require removing the content property and nesting that content as a child of the node ex.

Author: Updated

# Divider Spec

## Background

The divider control is intended to replace the Separator control currently exported. Its purpose is to visually separate content.
The purpose of redoing the component is to add minor structural updates and new engineering practices and update the naming conventions in the API to match naming conventions indicated by OpenUI research.

## Prior Art

[Github Epic Convergence Issue #16254](https://github.com/microsoft/fluentui/issues/16254)

[Github Convergence Issue #15759](https://github.com/microsoft/fluentui/issues/15759)

[Fluent UI Separator](https://developer.microsoft.com/en-us/fluentui#/controls/web/separator)

[Open UI Component Matrix](https://open-ui.org/analysis/component-matrix)

[Ant Design Divider](https://ant.design/components/divider/)

[Fast Divider](https://explore.fast.design/components/fast-divider)

[Semantic UI Divider](https://semantic-ui.com/elements/divider.html)

## Comparison of v8 and v0

- v8 - [Separator](https://developer.microsoft.com/en-us/fluentui#/controls/web/separator)

> - v0 - [Divider](https://fluentsite.z22.web.core.windows.net/0.51.4/components/divider/definition)

Author: Updated

## Sample Code

Basic Examples:

```html
<Divider />
<Divider vertical />
<Divider>This is a divider with content</Divider>
```

## Variants

The divider will have two main variants, rendering horizontally or vertically. The control is the same with the optional property `vertical` defining how it renders.

Standard default horizontal

```html
<Divider />
```

Vertical

```html
<Divider vertical />
```

Color

```html
<Divider color="black" />
```

Content

```html
<Divider>My Content</Divider>
```

Icon

```html
<Divider><Icon /></Divider>
```

Important

```html
<Divider important>This is important!</Divider>
```

Inset

```html
<Divider inset />
```

Appearance

```html
<Divider apperance="subtle" />
<Divider apperance="brand" />
<Divider apperance="strong" />
```

## API

> From [Divider.types.tsx](http://about:blank)

Author: Updated

### Slots

> - content (?)

Author: Updated

### Props

```ts
export interface DividerProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {

>   tokens?: RecursivePartial<DividerTokens>;

Author: Updated

  /**
   * Determines the alignment of the content within the divider.
   * @defaultvalue 'center'
   */
  alignContent?: 'start' | 'end' | 'center';

  appearance?: 'default' | 'subtle' | 'brand' | 'strong';

  children?: any;

  /* A divider can have a overriding border color */
  color?: string;

  /* A divider can be classified as important to emphasize its content */
  important?: boolean;

  inset?: boolean;

  /* A divider can be horizontal (default) or vertical*/
  vertical?: boolean;

  /* Overrides for custom appearances */
  height?: string;
  width?: string;

  fontSize?: string;
  fontWeight?: string;
  fontColor?: string;

  margin?: string;
  marginTop?: string;
  marginBottom?: string;
  marginLeft?: string;
  marginRight?: string;

  borderStyle?: string;

>   borderSize?: string | number;

Reviewer 3: I don't think that it's expected that we will include CSS properties to props interface. Some of them can be tweaked with component tokens, but this story is still in progress...

Is it intentional to have these props?

Author: This was tested and without them being defined, it creates a compile error as they are used / referenced in the useDividerStyles.tsx. I spoke with Mak and until there is a means of working around this, he mentioned to leave them.

Reviewer: I was telling @joschemd exactly that @layershifter, however, some of his customizations are based on them and we don't have a full story for component tokens. I'd say we can leave them now but possibly add a big comment noting that they will be removed once the component tokens story is done.

I also think that while leaving them in `Divider.types.ts` should be fine with such a comment we should remove them from the spec since it is just an interim step and not where we're intending to ultimately go.

Author: Updated

Border Color: tokens.alias.color.brand.brandBackgroundStatic

**Brand**

Font Color : tokens.alias.color.neutral.neutralForeground2

Border Color: tokens.alias.color.neutral.neutralStroke1

## Structure

```
<div>
  <!-- ::before to handle the divider line independent of the divider having content or not -->
  <div>Content</div>
  <!-- ::after to handle the divider line that goes after the content if it has been provided. -->
</div>
```

## Migration

The divider control is a direct replacement of the current Separator control. Notable changes are listed below:

Migration from V0 will require removing the content property and nesting that content as a child of the node ex.

```
<Divider content="My Content">
<Divider>My Content</Divider>

```

_Note:_ The property theme from the separator control has been removed

```
<Separator theme={myTheme}>
<Divider /> /* Handled by the theme provider */

```

_Note:_ The property styles from the separator control has been removed and now uses the inherent react style mechanics

```
<Separator styles={myStyles}>
<Divider style={{ direction:ltr }}/> /* Handled by the theme provider */

> ```

Reviewer: You should move this into a separate `MIGRATION.md` file and link it from here. You should also take a look at the format taken in the ones under [react-avatar](https://github.com/microsoft/fluentui/blob/master/packages/react-avatar/MIGRATION.md) and [react-link](https://github.com/microsoft/fluentui/blob/master/packages/react-link/MIGRATION.md).

Author: Updated
